test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -24,207 +24,266 @@
-{
-    sycl::queue q;
-    int n = 13;
-
-    // Allocate space for data using USM.
-    uint64_t* key_head = static_cast<uint64_t*>(sycl::malloc_shared(n * sizeof(uint64_t), q.get_device(), q.get_context()));
-    uint64_t* val_head = static_cast<uint64_t*>(sycl::malloc_shared(n * sizeof(uint64_t), q.get_device(), q.get_context()));
-    uint64_t* key_res_head = static_cast<uint64_t*>(sycl::malloc_shared(n * sizeof(uint64_t), q.get_device(), q.get_context()));
-    uint64_t* val_res_head = static_cast<uint64_t*>(sycl::malloc_shared(n * sizeof(uint64_t), q.get_device(), q.get_context()));
-
-    //T keys[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-    //T vals[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-
-    // keys_result = {1, 2, 3, 4, 1, 3, 1, 3, 0};
-    // vals_result = {1, 2, 3, 4, 2, 6, 2, 6, 0};
-
-    // Initialize data
-    for (int i = 0; i != 12; ++i) {
-        key_head[i] = i % 4 + 1;
-        val_head[i] = i % 4 + 1;
-        key_res_head[i] = 9;
-        val_res_head[i] = 1;
-        if (i > 3) {
-            ++i;
-            key_head[i] = key_head[i-1];
-            val_head[i] = val_head[i-1];
-            key_res_head[i] = 9;
-            val_res_head[i] = 1;
-        }
+    // specialization for non-random_access iterators
+    template <typename Policy, typename Iterator1, typename Iterator2, typename Iterator3, typename Iterator4,
+              typename Size>
+    typename ::std::enable_if<is_same_iterator_category<Iterator3, ::std::bidirectional_iterator_tag>::value ||
+                                  is_same_iterator_category<Iterator3, ::std::forward_iterator_tag>::value,
+                              void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last, Iterator2 vals_first, Iterator3 key_res,
+               Iterator4 val_res, Size n)
+    {
     }
-    key_head[12] = 0;
-    val_head[12] = 0;
-
-    // call algorithm
-    auto new_policy = oneapi::dpl::execution::make_device_policy<class reduce_by_segment_1>(q);
-    auto res1 = oneapi::dpl::reduce_by_segment(new_policy, key_head, key_head + n, val_head, key_res_head, val_res_head);
-
-    // check values
-    n = std::distance(key_res_head, res1.first);
-    for (auto i = 0; i != n; ++i) {
-        if (i < 4) {
-            ASSERT_EQUAL(key_res_head[i], i+1);
-            ASSERT_EQUAL(val_res_head[i], i+1);
-        } else if (i == 4 || i == 6) {
-            ASSERT_EQUAL(key_res_head[i], 1);
-            ASSERT_EQUAL(val_res_head[i], 2);
-        } else if (i == 5 || i == 7) {
-            ASSERT_EQUAL(key_res_head[i], 3);
-            ASSERT_EQUAL(val_res_head[i], 6);
-        } else if (i == 8) {
-            ASSERT_EQUAL(key_res_head[i], 0);
-            ASSERT_EQUAL(val_res_head[i], 0);
-        } else {
-            std::cout << "fail: unexpected values in output range\n";
+};
+
+#if !TEST_DPCPP_BACKEND_PRESENT
+const int max_n = 100000;
+#endif
+
+template <typename T, typename TestName>
+void
+test4buffers()
```

Comment:
I've done this refactoring. Please take a look.

Reply:
Device part looks good, but we need to test standard policies too. I suggest adding wrapper for it. It may look like `test` function in `all_of.pass.cpp` test, but with `invoke_on_all_host_policies` call inside. 

Just for note: we should improve our testing framework to avoid duplication in testing for standard and DPC++ policies. The existing tests which cover both types of policies do not cover USM cases, so they do not have such duplication.

Reply:
Added `test_on_host` function to invoke host policies.

Reply:
I've added a function that's invoked from main that calls `invoke_on_all_host_policies` as suggested.  Thanks.